File: 8332-168-983-minimum-cost-for-tickets/8332-168-983-minimum-cost-for-tickets.py
class Solution:
    def mincostTickets(self, days: list[int], costs: list[int]) -> int:
        lst = days[-1]
        dp = [0] * (lst + 1)

        day_set = set(days) 

        for d in range(1, lst + 1):
            if d not in day_set:
                dp[d] = dp[d - 1]
                continue
            ch1 = costs[0] + dp[d - 1]
            ch2 = costs[1] + dp[max(0, d - 7)]
            ch3 = costs[2] + dp[max(0, d - 30)]
            dp[d] = min(ch1, ch2, ch3)

        return dp[lst]

File: 8332-168-983-minimum-cost-for-tickets/8332-168-983-minimum-cost-for-tickets.java
import java.util.*;

class Solution {
    public int mincostTickets(int[] days, int[] costs) {
        int lst = days[days.length - 1];
        int[] dp = new int[lst + 1];

        Set<Integer> daySet = new HashSet<>();
        for (int day : days) {
            daySet.add(day);
        }

        for (int d = 1; d <= lst; d++) {
            if (!daySet.contains(d)) {
                dp[d] = dp[d - 1];
                continue;
            }
            int ch1 = costs[0] + dp[d - 1];
            int ch2 = costs[1] + dp[Math.max(0, d - 7)];
            int ch3 = costs[2] + dp[Math.max(0, d - 30)];
            dp[d] = Math.min(ch1, Math.min(ch2, ch3));
        }

        return dp[lst];
    }
}
